### PR TITLE
MAINT: add support for a==0 and x>0 edge case to igam and igamc

### LINKED ISSUE #10948: gammainc(0,x) = nan but should be 1, gammaincc(0,x) = nan but should be 0
The regularized incomplete gamma functions implemented in scipy return nan when the first parameter (a) is zero. Although a=0 is outside of the domain specified in the documentation, this domain restriction excluding a=0 is itself the bug: although the gamma function has a pole at 0, and the non-regularized incomplete gamma functions evaluated at a=0 give exponential integrals, the regularized versions are perfectly well behaved at a=0, see for example equation (2.1) in

http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.24.6764

Implementing a proper domain for these functions is important for example to fully capture the relationship between these functions and the CDF for the Poisson distribution. 

It looks like perhaps this behavior was copied over from boost, although I have not tested the boost implementation. The gsl documentation also specifies this a>0 domain for their regularized gamma functions so it may be a problem there too, although the version ported to ROOT is implemented for all real a, so I suspect gsl gives the correct values and just has the wrong documentation. Mathematica doesn't mention a domain restriction for these functions in their documentation, and Wolfram Alpha returns the expected values at a=0, so it looks properly implemented there. Anyway this is a bug report for scipy and I'm hoping at least for the problem to be fixed in scipy.

To reproduce the bug, do:
```
from scipy.special import gammainc, gammaincc
print(gammainc(0,1))
print(gammaincc(0,1))
```
This outputs
```
nan
nan
```
but the output should be
```
1
0
```

#### Scipy/Numpy/Python version information:
```
1.3.1 1.17.2 sys.version_info(major=3, minor=7, micro=4, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by rlucas7**
#### Reference issue
Closes #10948 

#### What does this implement/fix?
Adds the edge case of `a==0 && x>0` to not emit  `nan` in `igamc` and `igam` and adds unit tests for the new values.

### CODE REVIEW

File: scipy/special/cephes/igam.c [hunk 1]
```diff
@@ -129,9 +129,15 @@ double igam(double a, double x)
 {
     double absxma_a;
 
-    if (x < 0 || a <= 0) {
+    if (x < 0 || a < 0) {
 	sf_error("gammainc", SF_ERROR_DOMAIN, NULL);
 	return NPY_NAN;
+    } else if (a == 0) {
```
Comment (person142) [inline scipy/special/cephes/igam.c]: Let's also move up the `if (x == 0)` branch from below to keep all the various-parameters-zero logic clustered together.

File: scipy/special/cephes/igam.c [hunk 2]
```diff
@@ -129,9 +129,15 @@ double igam(double a, double x)
...
+	    if(x > 0) {
```
Comment (person142) [inline scipy/special/cephes/igam.c]: Space between the `if` and the `(`.

File: scipy/special/cephes/igam.c [hunk 3]
```diff
@@ -164,9 +170,15 @@ double igamc(double a, double x)
 {
     double absxma_a;
 
-    if (x < 0 || a <= 0) {
+    if (x < 0 || a < 0) {
 	sf_error("gammaincc", SF_ERROR_DOMAIN, NULL);
 	return NPY_NAN;
+    } else if (a == 0) {
+        if ( x > 0 ) {
```
Comment (person142) [inline scipy/special/cephes/igam.c]: Sub `(x > 0)` for `( x > 0 )`.

File: scipy/special/cephes/igam.c [hunk 4]
```diff
@@ -164,9 +170,15 @@ double igamc(double a, double x)
 {
     double absxma_a;
 
-    if (x < 0 || a <= 0) {
+    if (x < 0 || a < 0) {
 	sf_error("gammaincc", SF_ERROR_DOMAIN, NULL);
 	return NPY_NAN;
+    } else if (a == 0) {
```
Comment (person142) [inline scipy/special/cephes/igam.c]: As above about moving the `x == 0` branch up.

File: scipy/special/cephes/igam.c [hunk 5]
```diff
@@ -164,9 +170,15 @@ double igamc(double a, double x)
...
+        if ( x > 0 ) {
+		return 0;
+	} else {
+		return NPY_NAN;
```
Comment (person142) [inline scipy/special/cephes/igam.c]: Looks like your spacing settings are off-we want to indent by 4 spaces.

**Review by person142 — COMMENTED** [review]
Thanks for picking this one up @rlucas7. I would recommend also adding the following tests:

- Check that you get `nan` at `a = 0, x = 0`.
- Add a "limit check test", e.g. that `sc.gammaincc(1e-10, 1)` converges to the special case where `a = 0` and `x = 1`.

File: scipy/special/cephes/igam.c (hunk 1 above)
Comment (rlucas7) [reply scipy/special/cephes/igam.c]: done

File: scipy/special/cephes/igam.c (hunk 2 above)
Comment (rlucas7) [reply scipy/special/cephes/igam.c]: done, now reads `if (x > 0) {`

File: scipy/special/cephes/igam.c (hunk 3 above)
Comment (rlucas7) [reply scipy/special/cephes/igam.c]: done

File: scipy/special/cephes/igam.c (hunk 4 above)
Comment (rlucas7) [reply scipy/special/cephes/igam.c]: done

File: scipy/special/cephes/igam.c (hunk 5 above)
Comment (rlucas7) [reply scipy/special/cephes/igam.c]: not sure how the tabs got in there, fixed

**Comment by rlucas7** [comment]
> Thanks for picking this one up @rlucas7. I would recommend also adding the following tests:
> 
>     * Check that you get `nan` at `a = 0, x = 0`.
> 
>     * Add a "limit check test", e.g. that `sc.gammaincc(1e-10, 1)` converges to the special case where `a = 0` and `x = 1`.

I've added tests for both `sc.gammainc()` and `sc.gammaincc()` for limit check and for `nan` case so should be good on that end. Also, updated the style/formatting changes. Thanks for taking a look @person142  let's see if these changes pass on CI, they passed locally.

File: scipy/special/tests/test_gammainc.py [hunk 6]
```diff
@@ -53,6 +55,13 @@ def test_x_zero(self):
         a = np.arange(1, 10)
         assert_array_equal(sc.gammainc(a, 0), 0)
 
+    def test_nan_at_a_and_x_0(self):
+        assert np.isnan(sc.gammainc(0,0))
```
Comment (person142) [inline scipy/special/tests/test_gammainc.py]: You should be able to just add `(0, 0)` to the `INVALID_POINTS` list above.

File: scipy/special/tests/test_gammainc.py [hunk 7]
```diff
@@ -53,6 +55,13 @@ def test_x_zero(self):
...
+
+    def test_limit_check(self):
+        result = sc.gammainc(1e-10, 1)
```
Comment (person142) [inline scipy/special/tests/test_gammainc.py]: I would assert that `sc.gammainc(1e-10, 1)` is close to `sc.gammainc(0, 1)`.

File: scipy/special/tests/test_gammainc.py [hunk 8]
```diff
@@ -111,6 +123,14 @@ def test_infinite_limits(self):
             rtol=0
         )
 
+    def test_nan_a_and_x_0(self):
+        result = sc.gammaincc(0,0)
+        assert np.isnan(result)
+
+    def test_limit_check(self):
+        limit = sc.gammaincc(1e-10,1)
+        assert np.isclose(limit, 0)
```
Comment (person142) [inline scipy/special/tests/test_gammainc.py]: Same comment as above.

File: scipy/special/tests/test_gammainc.py [hunk 9]
```diff
@@ -111,6 +123,14 @@ def test_infinite_limits(self):
             rtol=0
         )
 
+    def test_nan_a_and_x_0(self):
```
Comment (person142) [inline scipy/special/tests/test_gammainc.py]: Same comment as above.

File: scipy/special/cephes/igam.c [hunk 10]
```diff
@@ -129,19 +129,25 @@ double igam(double a, double x)
...
 
-    if (x < 0 || a <= 0) {
-	sf_error("gammainc", SF_ERROR_DOMAIN, NULL);
-	return NPY_NAN;
-    } else if (npy_isinf(a)) {
-	if (npy_isinf(x)) {
-	    return NPY_NAN;
-	}
-	return 0;
+    if(x < 0 || a < 0) {
```
Comment (person142) [inline scipy/special/cephes/igam.c]: Space after the `if`.

**Review by person142 — COMMENTED** [review]
Just had a few small additional comments.

File: scipy/special/tests/test_gammainc.py (hunk 6 above)
Comment (rlucas7) [reply scipy/special/tests/test_gammainc.py]: done

File: scipy/special/tests/test_gammainc.py (hunk 7 above)
Comment (rlucas7) [reply scipy/special/tests/test_gammainc.py]: done.

File: scipy/special/tests/test_gammainc.py (hunk 8 above)
Comment (rlucas7) [reply scipy/special/tests/test_gammainc.py]: done

File: scipy/special/tests/test_gammainc.py (hunk 9 above)
Comment (rlucas7) [reply scipy/special/tests/test_gammainc.py]: same response as above, no need to test twice :)

File: scipy/special/cephes/igam.c (hunk 10 above)
Comment (rlucas7) [reply scipy/special/cephes/igam.c]: done

**Comment by rlucas7** [comment]
@person142 the 2 failing builds are unrelated, the travis builds fail because of failing tests in linear program routines which haven't been touched by this PR.

**Comment by person142** [comment]
> the 2 failing builds are unrelated

Reran them and they passed on the second try.

**Pull request merged**